@uppy/aws-s3 rewrite blog (#429)
transloadit/uppy.io: blog/2026-03-18-aws-s3-rewrite.md

replaces 11 overlapping callbacks (which conflated signing with S3 protocol
details) with 3 mutually exclusive signing modes plus a few orthogonal
config options. Companion is now optional: you can talk directly to any
S3-compatible service from the browser.

<!--truncate-->

## Why a rewrite?

The old plugin was the product of years of incremental additions. What started
as a multipart uploader grew to handle single-part PUT uploads, presigned POST
uploads, client-side SigV4 signing, Companion-backed signing, and more all in
one class with shared state and interleaved code paths. The
[tracking issue](https://github.com/transloadit/uppy/issues/6055) sums it up: it
had become completely unwieldy and was the plugin with the
[most reported problems](https://github.com/transloadit/uppy/issues?q=is%3Aissue%20state%3Aopen%20aws%20s3).

The new architecture splits things into two layers:

- **S3mini**: a standalone, browser-native S3 client with zero Uppy
  dependencies. It handles presigned URLs, SigV4 signing, multipart
  orchestration, XML parsing, retries with exponential backoff, offline
  detection, and credential caching. It uses the Web Crypto API and

>   `XMLHttpRequest`. No AWS SDK needed.

Reviewer: maybe clarify why we don't use `fetch` which is more modern

- **The plugin**: a thin wrapper that wires S3mini to Uppy’s file lifecycle
  (progress events, pause/resume, abort, Golden Retriever state).

Both sit on top of an `S3Client` abstract base, with `S3mini` and
`CompanionS3` as the two concrete implementations. The orchestration code
above doesn’t care which one is in use, so progress, retries, pause/resume,
and resume-from-refresh all behave identically across signing modes.

Each layer has one job. S3 protocol concerns stay in the client, Uppy concerns
stay in the plugin.

The S3mini client is forked from
[good-lly/s3mini](https://github.com/good-lly/s3mini) (MIT-licensed, by Jølly

> Good), modified for browser use and integrated with Uppy’s upload lifecycle.

Reviewer: ```suggestion
Good), simplified and streamlined for browser use and integrated with Uppy’s upload lifecycle.
```

		});
		return res.json(); // { url: 'https://...presigned-url' }
	},
});
```

### `companionEndpoint` : Companion signing

Point the plugin at your Companion server. Nothing else to configure.

```js
new Uppy().use(AwsS3, {
	companionEndpoint: 'https://companion.example.com',
});
```

Options like `shouldUseMultipart`, `getChunkSize`, `limit`, `generateObjectKey`,
and `allowedMetaFields` are still available as simple configuration. They’re
just not callbacks that replace S3 operations anymore.

## Companion is now optional

The old plugin essentially required a Companion server (or a custom callback
for every S3 operation) to do anything. The new plugin lets you talk directly
to any S3-compatible service from the browser using `signRequest` or
`getCredentials`. No Companion in the data path.

Existing Companion deployments still work unchanged. The `companionEndpoint`
mode reuses the same Companion endpoints the old plugin used, so server-side
key generation (`config.getKey()`), STS credential issuance, and presigning
all behave the same. The only migration step on the client is renaming
`endpoint` → `companionEndpoint`.

> What changed under the hood: the client is now layered on top of an `S3Client`

Reviewer: i believe this was mentioned also above?

infrastructure means contributors can reproduce signing bugs locally without
hitting AWS.

## Migration guide

### From `endpoint` (Companion) mode

```js
// Before
new Uppy().use(AwsS3, {
	endpoint: 'https://companion.example.com',
});

// After
new Uppy().use(AwsS3, {
	companionEndpoint: 'https://companion.example.com',
});
```

### From `getTemporarySecurityCredentials`

```js
// Before
new Uppy().use(AwsS3, {
	endpoint: 'https://companion.example.com',
	getTemporarySecurityCredentials: true,
});

// After
new Uppy().use(AwsS3, {
	s3Endpoint: 'https://my-bucket.s3.us-east-1.amazonaws.com',
	region: 'us-east-1',
	getCredentials: async () => {
		const res = await fetch('/api/s3/sts');
		return res.json();
	},
});
```

### From custom callbacks

```js
// Before: 6 separate callbacks
new Uppy().use(AwsS3, {
	getUploadParameters: (file) => {
		/* ... */
	},
	createMultipartUpload: (file) => {
		/* ... */
	},
	signPart: (file, partData) => {
		/* ... */
	},
	completeMultipartUpload: (file, data) => {
		/* ... */
	},
	abortMultipartUpload: (file, data) => {
		/* ... */
	},
	listParts: (file, data) => {
		/* ... */
	},
});

> // After: one callback

Reviewer: TBH I find this new behavior more confusing than the previous behavior. We should probably document it

---
title: 'Rewriting @uppy/aws-s3 from scratch'
date: 2026-03-18
authors: [prakash, mifi, murderlon]
slug: 'aws-s3-rewrite'
published: false
toc_max_heading_level: 2
---

`@uppy/aws-s3` has been Uppy’s most-used uploader plugin, but also the one
with the most bug reports. Over the years, it accumulated nearly 2,000 lines
of tightly coupled code, 11 user-facing callback options, and 8 Companion
endpoints. Every bug fix risked breaking something else and nobody wanted to
touch it.

So we rewrote it. 

The new plugin closes
[**13 open issues**](https://github.com/transloadit/uppy/issues/6055) and
replaces 11 overlapping callbacks (which conflated signing with S3 protocol
details) with 3 mutually exclusive signing modes, plus a few orthogonal config
options. Companion is now optional: you can talk directly to any S3-compatible
service from the browser.

<!--truncate-->

## Why a rewrite?

The old plugin was the product of years of incremental additions. What started
as a multipart uploader grew to handle single-part PUT uploads, presigned POST

> uploads, client-side SigV4 signing, Companion-backed signing, and more – all in

Reviewer 2: Not a 100% sure I interpreted this sentence correctly. It seemed to me the last part needed to be separated.

Author: I agree, readability can be improved, can we split it into different sentences like this ?
>Companion-backed signing, and more. All of it lived in one class with shared state and interleaved code paths.

what do you think ?

Reviewer 2: Yep! That works as well. Looks good to me 👍

> uploads, client-side SigV4 signing, Companion-backed signing, and more – all in

Reviewer 3: Unfortunately LLMs ruined em-dashes. In this case, we can turn it into a comma.

Reviewer 2: Technically, it's an en-dash. But I regrettably see your point. Unfortunate that one of my favorite punctuation tools has become suspicious. No matter, though, since it's already changed to something else!

one class with shared state and interleaved code paths. The
[tracking issue](https://github.com/transloadit/uppy/issues/6055) sums it up: the
plugin had become completely unwieldy, accumulating the highest number of
[reported problems](https://github.com/transloadit/uppy/issues?q=is%3Aissue%20state%3Aopen%20aws%20s3).

The new architecture splits things into two layers:

- **S3mini**: a standalone, browser-native S3 client that doesn’t depend on
  Uppy’s plugin system (the only Uppy code it uses is the internal `fetcher`
  helper for XHR with retries). It handles presigned URLs, SigV4 signing,
  multipart orchestration, XML parsing, retries with exponential backoff,
  offline detection, and credential caching. It uses the Web Crypto API and
  `XMLHttpRequest`. (`XMLHttpRequest` is used instead of `fetch`, because the
  Fetch API does not expose upload progress events; XHR’s
  `xhr.upload.onprogress` is the only standard way to stream byte-count updates

>   into Uppy’s progress system.) No AWS SDK needed.

Reviewer 3: _“Web Crypto API”_ could link to https://developer.mozilla.org/docs/Web/API/Web_Crypto_API.

I don’t think we need to explain why we use `XMLHttpRequest` instead of `fetch`. This is an implementation detail.

_“No AWS SDK needed.”_ is not a full sentence. It needs a verb: _“No AWS SDK is needed.”_

Reviewer 2: True, true. Can't go wrong with abiding by the rules. Leaving out the verb here didn't bother me. A bit more informal and, to me, it fits the rhythm.

Author: > I don’t think we need to explain why we use XMLHttpRequest instead of fetch. This is an implementation detail.

It was added in response to this https://github.com/transloadit/uppy.io/pull/429#discussion_r3207458240 , I think it makes sense to keep this.

Your backend returns temporary STS credentials. The plugin signs all requests in
the browser using SigV4. No server round-trip per request.

```js
new Uppy().use(AwsS3, {
	s3Endpoint: 'https://my-bucket.s3.us-east-1.amazonaws.com',
	getCredentials: async () => {
		const res = await fetch('/api/s3/credentials');
		return res.json();
		// {
		//   credentials: { accessKeyId, secretAccessKey, sessionToken },
		//   region: 'us-east-1',
		// }
	},
});
```

The `region` used for signing comes from the `getCredentials` response. You can
also pass it as a plugin option. If neither is set, it falls back to `auto`. This
is fine for region-less services like Cloudflare R2, but for AWS you need to
supply the real region one way or the other.

### `signRequest` : bring your own signer

Your backend signs each request and returns a presigned URL. No other options
needed; the upload location is derived from the presigned URL itself.

```js
new Uppy().use(AwsS3, {
	signRequest: async ({ method, key, uploadId, partNumber }) => {
		const res = await fetch('/api/s3/sign', {
			method: 'POST',
			body: JSON.stringify({ method, key, uploadId, partNumber }),
		});

> 		return res.json(); // { url: 'https://...presigned-url' }

Reviewer 3: The HTTP spec has reserved the `.example` TLD for examples.

```suggestion
		return res.json(); // { url: 'https://presigned-url.example' }
```

	},
});
```

### `companionEndpoint` : Companion signing

Point the plugin at your Companion server. Nothing else to configure.

```js
new Uppy().use(AwsS3, {

> 	companionEndpoint: 'https://companion.example.com',

Reviewer 3: The HTTP spec has reserved the `.example` TLD and `example.com` for examples. I’m not sure about `example.com` subdomains.

```suggestion
	companionEndpoint: 'https://companion.example',
```
